kernel/src/actions/deletion_vector_writer.rs

```diff
@@ -0,0 +1,733 @@
+/// - Followed by a 4-byte magic number (0x64485871, little-endian)
+/// - Followed by the serialized 64-bit Roaring Bitmap
+/// - Followed by a 4-byte CRC32 checksum (big-endian) of the serialized data
+///
+/// # Examples
+///
+/// ```rust,ignore
+/// use std::io::Cursor;
+/// use delta_kernel::actions::deletion_vector_writer::{StreamingDeletionVectorWriter, KernelDeletionVector};
+///
+/// let mut buffer = Vec::new();
+/// let mut writer = StreamingDeletionVectorWriter::new(&mut buffer);
+///
+/// let mut dv = KernelDeletionVector::new();
+/// dv.add_deleted_row_indexes([1, 5, 10].iter().copied());
+///
+/// let descriptor = writer.write_deletion_vector(dv)?;
+/// writer.finalize()?;
+/// # Ok::<(), delta_kernel::Error>(())
+/// ```
+pub struct StreamingDeletionVectorWriter<'a, W: Write> {
+    writer: &'a mut W,
+    current_offset: i32,
+    has_written_version: bool,
+}
+
+impl<'a, W: Write> StreamingDeletionVectorWriter<'a, W> {
+    /// Create a new streaming deletion vector writer.
+    ///
+    /// # Arguments
+    ///
+    /// * `writer` - A mutable reference to any type implementing [`std::io::Write`].
+    ///
+    /// # Examples
+    ///
+    /// ```rust,ignore
+    /// let mut buffer = Vec::new();
+    /// let writer = StreamingDeletionVectorWriter::new(&mut buffer);
+    /// ```
+    pub fn new(writer: &'a mut W) -> Self {
+        Self {
+            writer,
+            current_offset: 0,
+            has_written_version: false,
+        }
+    }
+
+    /// Write a deletion vector to the underlying writer.
+    ///
+    /// This method can be called multiple times to write multiple deletion vectors to the same
+    /// writer. The caller is responsible for keeping track of which deletion vector corresponds to
+    /// which data file.
+    ///
+    /// # Arguments
+    ///
+    /// * `deletion_vector` - The deletion vector to write
+    ///
+    /// # Returns
+    ///
+    /// A [`DeletionVectorWriteResult`] containing the offset, size, and cardinality
+    /// of the written deletion vector.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - The writer fails to write data
+    /// - The deletion vector cannot be serialized
+    /// - The offset or size would overflow an i32
+    ///
+    /// # Examples
+    ///
+    /// ```rust,ignore
+    /// let mut dv = KernelDeletionVector::new();
+    /// dv.add_deleted_row_indexes([1, 5, 10].iter().copied());
+    ///
+    /// let descriptor = writer.write_deletion_vector(dv)?;
+    /// println!("Written DV at offset {} with size {}", descriptor.offset, descriptor.size_in_bytes);
+    /// # Ok::<(), delta_kernel::Error>(())
+    /// ```
+    pub fn write_deletion_vector(
+        &mut self,
+        deletion_vector: impl DeletionVector,
+    ) -> DeltaResult<DeletionVectorWriteResult> {
+        // Write version byte on first write
+        if !self.has_written_version {
+            self.writer
```

Comment:
```suggestion
        if !self.has_written_version {
            debug_assert!(self.current_offset == 0);
            self.writer
```

Reply:
removed has_written_version, to just use current_offset.